pkg/kstatus/watcher/watcher_dynamic.go

```diff
@@ -0,0 +1,825 @@
+	id object.ObjMetadata,
+) context.CancelFunc {
+	// schedule delayed event
+	klog.V(4).Infof("ScheduleWindow (%v) timer started: object unschedulable: %s", status.ScheduleWindow, id)
+	delayCtx, cancelFunc := context.WithDeadline(context.Background(),
+		time.Now().Add(status.ScheduleWindow))
+	go func() {
+		select {
+		case <-ctx.Done():
+			// stop waiting
+			cancelFunc()
+		case <-delayCtx.Done():
+			if delayCtx.Err() == context.DeadlineExceeded {
+				klog.V(4).Infof("ScheduleWindow (%v) elapsed without status update: re-reading object status: %s", status.ScheduleWindow, id)
+				// check again
+				rs, err := w.readStatusFromCluster(ctx, id)
+				if err != nil {
+					handleError(eventCh, err)
+					return
+				}
+				eventCh <- event.Event{
+					Type:     event.ResourceUpdateEvent,
+					Resource: rs,
+				}
+			}
+			// else stop waiting
+		}
+	}()
+	return cancelFunc
+}
+
+func handleError(eventCh chan<- event.Event, err error) {
+	eventCh <- event.Event{
+		Type:  event.ErrorEvent,
+		Error: err,
+	}
+}
+
+// watchErrorHandler logs errors and cancels the informer for this GroupKind
+// if the NotFound error is received, which usually means the CRD was deleted.
+// Based on DefaultWatchErrorHandler from k8s.io/client-go@v0.23.2/tools/cache/reflector.go
+func (w *ObjectStatusInformer) watchErrorHandler(groupKind schema.GroupKind, err error) {
+	switch {
+	case err == io.EOF:
+		// Watch closed normally
+	case err == io.ErrUnexpectedEOF:
+		// Keep retrying
+		klog.V(1).Infof("Watch of %v closed: unexpected EOF: %v", groupKind, err)
+	case apierrors.IsResourceExpired(err): // resourceVersion too old
+		// Keep retrying
+		klog.V(4).Infof("Watch of %v closed: %v", groupKind, err)
+	case apierrors.IsGone(err): // DEPRECATED
+		// Keep retrying
+		klog.V(4).Infof("Watch of %v closed: %v", groupKind, err)
+	case apierrors.IsNotFound(err) || containsNotFoundMessage(err): // CRD deleted or not created
+		// Stop watching this resource
+		klog.V(3).Infof("Watch for %s errored: stopping watcher: %v", groupKind, err)
+		// Add the GroupKind to the pendingWatchers,
+		// so if the CRD is re-applied, the informer can be re-created.
+		w.pendingWatchers.Put(groupKind)
+		// Cancel the current informer for this GroupKind.
+		w.watchCanceller.Cancel(groupKind)
+	default:
+		// Keep retrying
+		klog.Errorf("Error watching %v: %v", groupKind, err)
+	}
+}
+
+const resourceNotFoundMessage = "the server could not find the requested resource"
+
+// containsNotFoundMessage checks if the error string contains the message for
+// StatusReasonNotFound.
+// See k8s.io/apimachinery@v0.23.2/pkg/api/errors/errors.go
+// This is necessary because the Informer doesn't properly wrap errors.
+func containsNotFoundMessage(err error) bool {
+	return strings.Contains(err.Error(), resourceNotFoundMessage)
+}
+
+// uniqueGKs returns a set of unique GroupKinds from a set of object identifiers.
+func uniqueGKs(ids object.ObjMetadataSet) []schema.GroupKind {
```

Comment:
If this function returns `[]schema.GroupVersionKind`, then in each watcher, you don't need to look up the GVK from the restmapper. You can also get the `pendingWatchers` from it.

Reply:
The uniqueness is by GroupKind tho, we wouldn't want to allow multiple Versions of the same GK to be watched simultaneously.